### PR TITLE
Select query

### PR DESCRIPTION
**Pull request by evasomething**

### CODE REVIEW

File: src/client/components/cube-view/cube-view.tsx [hunk 1]
```diff
@@ -258,6 +267,26 @@ export class CubeView extends React.Component<CubeViewProps, CubeViewState> {
...
+  onRawDataModalClose() {
+    this.setState({
+      showRawDataModal: false
+    });
+  }
+
+  renderRawDataModal() {
+    const { RawDataModalAsync, showRawDataModal, essence, visualizationStage } = this.state;
+    if (!RawDataModalAsync || !showRawDataModal) return null;
+    return <RawDataModalAsync stage={visualizationStage}
```
Comment (vogievetsky) [inline src/client/components/cube-view/cube-view.tsx]: Please indent so stage is on its own line

File: package.json [hunk 2]
```diff
@@ -24,6 +24,7 @@
   },
   "dependencies": {
     "body-parser": "1.15.0",
+    "browser-filesaver": "^1.1.1",
```
Comment (vogievetsky) [inline package.json]: remove ^ please

File: src/client/components/raw-data-modal/raw-data-modal.tsx [hunk 3]
```diff
@@ -0,0 +1,312 @@
+import {Manifest} from "../../../common/models/manifest/manifest";
```
Comment (vogievetsky) [inline src/client/components/raw-data-modal/raw-data-modal.tsx]: I know this was not you but WebStorm but please watch out with the auto generated imports they tend to be wild. This should go in line 7

File: src/client/components/raw-data-modal/raw-data-modal.tsx [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    const dimensionsWidth = this.dimensionsWidths.reduce((pV: number, cV: number) => {
+      return pV + cV;
+    }, 0);
+    const measuresWidths = this.measuresWidths.reduce((pV: number, cV: number) => {
+      return pV + cV;
+    }, 0);
+    const rowWidth = dimensionsWidth + measuresWidths;
+    const title = `${makeTitle(SEGMENT.toLowerCase())} ${STRINGS.rawData} `;
+    const dataLength = dataset ? dataset.data.length : 0;
+    const bodyHeight = dataLength * SimpleTable.ROW_HEIGHT;
```
Comment (vogievetsky) [inline src/client/components/raw-data-modal/raw-data-modal.tsx]: ROW_HEIGHT should instead be a parameter to SimpleTable

File: src/client/components/filter-tile/filter-tile.tsx [hunk 5]
```diff
@@ -27,21 +28,36 @@ export interface ItemBlank {
...
 
-function formatLabel(dimension: Dimension, clause: FilterClause, essence: Essence): string {
+export interface LabelFormatOptions {
+  dimension: Dimension;
+  clause: FilterClause;
+  essence: Essence;
+  verbose?: boolean;
+}
+
+export function formatLabel(options: LabelFormatOptions): string {
```
Comment (vogievetsky) [inline src/client/components/filter-tile/filter-tile.tsx]: rather than exporting a function from a component it should be moved into a shared util

File: src/client/components/scroll-container/scroll-container.tsx [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+  render() {
+    const { style, onScroll, onMouseLeave, onMouseMove, onClick } = this.props;
+    return <div
+      className="scroll-container"
+      ref="base"
+      onScroll={onScroll.bind(this)}
+      onMouseLeave={onMouseLeave ? onMouseLeave.bind(this) : null}
+      onMouseMove={onMouseMove ? onMouseMove.bind(this) : null}
+      onClick={onClick ? onClick.bind(this) : null}
```
Comment (vogievetsky) [inline src/client/components/scroll-container/scroll-container.tsx]: you do not need to bind(this) if you are just passing them in

File: src/client/components/raw-data-modal/raw-data-modal.tsx [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+      this.dimensionsWidths = this.dimensionsWidths.concat(width);
+      cols = cols.concat(<div className={thClassName} style={style} key={i}>
+        <div className={thText}>
+          {key}
+        </div>
+      </div>);
+    });
+
+    this.measuresWidths = [];
+    measures.map((measure, i) => {
```
Comment (vogievetsky) [inline src/client/components/raw-data-modal/raw-data-modal.tsx]: this is fundamentally wrong. This dialog is doing a raw (select) query. measures and dimensions do not exist here. only attributes (columns). For dimensions we logically could add the derived dimensions to the query (only once Druid goes to 0.9.1 though) but measures only exist when there is a split. The string "measure" should not exist in this file.

File: src/client/components/raw-data-modal/raw-data-modal.tsx [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+    return `${dsName}-${visType}-${wordsOnly}`;
+  }
+
+
+  render() {
+    const { onClose, stage } = this.props;
+    const { dataset, loading, scrollTop, scrollLeft, error } = this.state;
+    const headerColumns = this.renderHeader();
+    const rows = this.renderRows(dataset, scrollTop, stage);
+    const dimensionsWidth = this.dimensionsWidths.reduce((pV: number, cV: number) => {
```
Comment (vogievetsky) [inline src/client/components/raw-data-modal/raw-data-modal.tsx]: omg! make a `sum` util already :-p

File: src/client/components/raw-data-modal/raw-data-modal.tsx [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+      return pV + cV;
+    }, 0);
+    const measuresWidths = this.measuresWidths.reduce((pV: number, cV: number) => {
+      return pV + cV;
+    }, 0);
+    const rowWidth = dimensionsWidth + measuresWidths;
+    const title = `${makeTitle(SEGMENT.toLowerCase())} ${STRINGS.rawData} `;
```
Comment (vogievetsky) [inline src/client/components/raw-data-modal/raw-data-modal.tsx]: lol trailing space in a string ?

File: src/client/components/scroll-container/scroll-container.tsx [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+export interface ScrollContainerState {
+}
+
+export class ScrollContainer extends React.Component<ScrollContainerProps, ScrollContainerState> {
+
+  render() {
+    const { style, onScroll, onMouseLeave, onMouseMove, onClick } = this.props;
+    return <div
+      className="scroll-container"
```
Comment (vogievetsky) [inline src/client/components/scroll-container/scroll-container.tsx]: I think the outer component should be called "scroller" and the inner div should be "scroller-inner" or something

File: src/client/components/scroll-container/scroll-container.tsx [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+import { $, Expression, Executor, Dataset } from 'plywood';
+import { Stage, Clicker, Essence, DataSource, Filter, Dimension, Measure } from '../../../common/models/index';
+// import { ... } from '../../config/constants';
+import { SvgIcon } from '../svg-icon/svg-icon';
+import { Fn } from "../../../common/utils/general/general";
+
+export interface ScrollContainerProps extends React.Props<any> {
+  onScroll: Fn;
```
Comment (vogievetsky) [inline src/client/components/scroll-container/scroll-container.tsx]: You should wrap this function in scroller logic.

File: src/client/components/hiluk-menu/hiluk-menu.tsx [hunk 12]
```diff
@@ -91,7 +99,7 @@ export class HilukMenu extends React.Component<HilukMenuProps, HilukMenuState> {
       onClick={this.openRawDataModal.bind(this)}
     >{STRINGS.viewRawData}</li>);
 
-    if (getVisualizationDataset) {
+    if (getVisualizationDataset()) {
```
Comment (vogievetsky) [inline src/client/components/hiluk-menu/hiluk-menu.tsx]: why this change?

File: src/common/models/filter/filter.ts [hunk 13]
```diff
@@ -215,6 +215,14 @@ export class Filter implements Instance<FilterValue, FilterJS> {
     return clauses.get(index).getTimeRange();
   }
 
+  public getFileString(attribute: Expression) {
```
Comment (vogievetsky) [inline src/common/models/filter/filter.ts]: better name here for parameter?

File: src/common/models/filter/filter.ts [hunk 14]
```diff
@@ -215,6 +215,14 @@ export class Filter implements Instance<FilterValue, FilterJS> {
...
+    var timeRange = this.getTimeRange(attribute);
```
Comment (vogievetsky) [inline src/common/models/filter/filter.ts]: check that timeRange exists (it might not) there might not be a time filter

File: src/client/components/hiluk-menu/hiluk-menu.tsx (hunk 12 above)
Comment (evasomething) [inline src/client/components/hiluk-menu/hiluk-menu.tsx]: if there's no visualization doesn't it make sense for the export option to not show up?

**Comment by vogievetsky** [comment]
gr 🎱
